From 6ae90b1e9286f473269c79eb22087a42efd5daf4 Mon Sep 17 00:00:00 2001
From: Ruchi Kandoi <kandoiruchi@google.com>
Date: Fri, 1 Feb 2019 12:17:21 -0800
Subject: [PATCH 2/5] Prevent OOB error in rw_i93_sm_detect_ndef() am:
 9939edeb9f am: 2ee2f756d0 am: d4921e738b

Change-Id: I4199f067eac626944f60e931183a62eb1d6946cf
(cherry picked from commit 8909dbdece1c98e21c2190fe968013e0dc50f347)
---
 src/nfc/tags/rw_i93.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/src/nfc/tags/rw_i93.c b/src/nfc/tags/rw_i93.c
index acdbcb7..c3a2f26 100644
--- a/src/nfc/tags/rw_i93.c
+++ b/src/nfc/tags/rw_i93.c
@@ -23,6 +23,7 @@
  *  mode.
  *
  ******************************************************************************/
+#include <log/log.h>
 #include <string.h>
 #include "nfc_target.h"
 #include "bt_types.h"
@@ -1506,6 +1507,11 @@ void rw_i93_sm_detect_ndef (BT_HDR *p_resp)
     RW_TRACE_DEBUG1 ("rw_i93_sm_detect_ndef () sub_state:0x%x", p_i93->sub_state);
 #endif
 
+    if (length == 0) {
+        android_errorWriteLog(0x534e4554, "121260197");
+        rw_i93_handle_error(NFC_STATUS_FAILED);
+        return;
+    }
     STREAM_TO_UINT8 (flags, p);
     length--;
 
@@ -1529,6 +1535,11 @@ void rw_i93_sm_detect_ndef (BT_HDR *p_resp)
     {
     case RW_I93_SUBSTATE_WAIT_UID:
 
+        if (length < (I93_UID_BYTE_LEN + 1)) {
+            android_errorWriteLog(0x534e4554, "121260197");
+            rw_i93_handle_error(NFC_STATUS_FAILED);
+            return;
+        }
         STREAM_TO_UINT8 (u8, p); /* DSFID */
         p_uid = p_i93->uid;
         STREAM_TO_ARRAY8 (p_uid, p);
-- 
2.7.4

